community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeAllEntriesLabelScanReader.java

```diff
@@ -0,0 +1,258 @@
+{
+    private final IntFunction<RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException>> seekProvider;
+    private final List<RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException>> cursors = new ArrayList<>();
+    private final int highestLabelId;
+
+    NativeAllEntriesLabelScanReader( IntFunction<RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException>> seekProvider,
+            int highestLabelId )
+    {
+        this.seekProvider = seekProvider;
+        this.highestLabelId = highestLabelId;
+    }
+
+    @Override
+    public long maxCount()
+    {
+        return -1;
+    }
+
+    @Override
+    public Iterator<NodeLabelRange> iterator()
+    {
+        try
+        {
+            long lowestRange = Long.MAX_VALUE;
+            closeCursors();
+            for ( int labelId = 0; labelId <= highestLabelId; labelId++ )
+            {
+                RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor = seekProvider.apply( labelId );
+
+                // Bootstrap the cursor, which also provides a great opportunity to exclude if empty
+                if ( cursor.next() )
+                {
+                    lowestRange = min( lowestRange, cursor.get().key().idRange );
+                    cursors.add( cursor );
+                }
+            }
+            return new NodeLabelRangeIterator( lowestRange );
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
+        }
+    }
+
+    private void closeCursors() throws IOException
+    {
+        for ( RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor : cursors )
+        {
+            cursor.close();
+        }
+        cursors.clear();
+    }
+
+    @Override
+    public void close() throws Exception
+    {
+        closeCursors();
+    }
+
+    /**
+     * The main iterator over {@link NodeLabelRange ranges}, aggregating all the cursors as it goes.
+     */
+    private class NodeLabelRangeIterator extends PrefetchingIterator<NodeLabelRange>
+    {
+        private long currentRange;
+        private final ReusableNodeLabelRange current = new ReusableNodeLabelRange();
+
+        // nodeId (relative to lowestRange) --> labelId[]
+        @SuppressWarnings( "unchecked" )
+        private final List<Long>[] labelsForEachNode = new List[RANGE_SIZE];
+
+        NodeLabelRangeIterator( long lowestRange )
+        {
+            this.currentRange = lowestRange;
+        }
+
+        @Override
+        protected NodeLabelRange fetchNextOrNull()
+        {
+            if ( currentRange == Long.MAX_VALUE )
+            {
+                return null;
+            }
+
+            fill( labelsForEachNode, null );
+            long nextLowestRange = Long.MAX_VALUE;
+            int slots = 0;
+            try
+            {
+                // One "rangeSize" range at a time
+                for ( RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor : cursors )
```

Comment:
Don't think cursors are shared. Iterator is responsible for moving cursors (those that are on the current range) forwards and cram their range data into a NodeLabelRange, which is returned out from the iterator.

Although this comment reminded me about a fix we did after this PR branch, which is that the NodeLabelRange instances can be processed by multiple threads once they leave the iterator, so changed ReusableNodeLabelRange to not be reused, instead creating new instances every time.